src/components/FileBrowser.tsx

```diff
@@ -771,36 +657,89 @@ export default class FileBrowser extends React.Component<
       'chonky-fill-parent': fillParentContainer === true,
     });
     return (
-      <div ref={this.ref} className={className}>
-        <Controls
-          folderChain={folderChain}
-          selection={selection}
-          onFileOpen={onFileOpen}
-          onFolderCreate={onFolderCreate}
-          onUploadClick={onUploadClick}
-          onDownloadFiles={onDownloadFiles}
-          onDeleteFiles={onDeleteFiles}
-          getFilesFromSelection={this.getFilesFromSelection}
-          view={view}
-          setView={this.setView}
+      <ConfigContext.Provider
+        value={{ Icon, icons: { ...defaultIcons, ...icons } }}
+      >
+        <SelectionController
+          disableSelection={disableSelection}
+          files={sortedFiles}
+          folderChain={isNil(folderChain) ? [] : folderChain}
+          onSelectionChange={onSelectionChange}
           options={options}
-          setOption={this.setOption}
-          activateSortProperty={this.activateSortProperty}
           sortProperty={sortProperty}
           sortOrder={sortOrder}
-        />
-        <FileList
-          files={sortedFiles}
-          selection={selection}
-          doubleClickDelay={doubleClickDelay as number}
-          onFileSingleClick={this.handleFileSingleClick}
-          onFileDoubleClick={this.handleFileDoubleClick}
-          thumbnailGenerator={thumbnailGenerator}
-          showRelativeDates={options[Option.ShowRelativeDates]}
-          fillParentContainer={fillParentContainer === true}
-          view={view}
-        />
-      </div>
+        >
+          {({ selection = {}, handleFileSingleClick: selectionHandler }) => {
+            const handleFileSingleClick: InternalClickHandler = (
+              file: FileData,
+              displayIndex: number,
+              event: InputEvent
+            ) => {
```

Comment:
Correct me if I'm wrong:

It looks like every time we change the selection `handleFileSingleClick` get re-created, and then passed down to pretty much all children down the tree. Will that not cause serious performance regressions? Or is there some caching logic magically happening somewhere?

Reply:
There is no magic caching that I know of.  I'm wasn't worried about this because this component will soon change to be functional, and then we can `useCallback`

As it stands though, I'd be surprised if the performance overhead is all that much.  If I were planning to leave it, I'd run some performance comparisons, but since it's temporary, its probably not worth the time.

Reply:
Though I should mention, this is the thing I miss most about class components.

Reply:
if this inner render prop is turned into a function component react can bail out on rendering

Reply:
Sounds good.